### PR TITLE
Remove newly-invalidated records from synchronization

### PR DESCRIPTION
**Pull request by ahorner**
These changes introduce `Restforce::DB::Cleaner`, an object which coordinates the detection of "invalid" Salesforce records and removal of the corresponding database records from the system.

I've inserted the invocation of the `Cleaner` into the "Propagate" step, since cleaning up records logically fits with the process of record  propagation, but I could be argued into splitting it into its own discretely-logged step.

### CODE REVIEW

**Comment by rdunlop** [comment]
:+1:

File: lib/restforce/db/cleaner.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      end
+
+      # Internal: Get the IDs of all recently-modified Salesforce records
+      # corresponding to the object type for this mapping.
+      #
+      # Returns an Array of IDs.
+      def all_salesforce_ids
+        all_ids = []
+
+        @mapping.unscoped do |m|
```
Comment (danhodos) [inline lib/restforce/db/cleaner.rb]: I'd prefer if we could refrain from using one-letter variable names.

File: test/lib/restforce/db/cleaner_test.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
+require_relative "../../../test_helper"
+
+describe Restforce::DB::Cleaner do
+
+  configure!
+  mappings!
+
+  let(:cleaner) { Restforce::DB::Cleaner.new(mapping) }
+
+  describe "#run", vcr: { match_requests_on: [:method, VCR.request_matchers.uri_without_param(:q)] } do
```
Comment (danhodos) [inline test/lib/restforce/db/cleaner_test.rb]: I hadn't seen `match_requests_on` before. Neat!

File: lib/restforce/db/worker.rb [hunk 3]
```diff
@@ -128,6 +128,7 @@ def runner
       # Returns nothing.
       def propagate(mapping)
         Initializer.new(mapping, runner).run
+        Cleaner.new(mapping, runner).run
```
Comment (danhodos) [inline lib/restforce/db/worker.rb]: Yeah, I'd say go ahead and move this into its own method and log it separately.

**Comment by danhodos** [comment]
Word word, had some minor comments, but nothing major. I kinda wish we didn't need to do two separate API requests for this cleaning step, but I'm for sure not about to worry about performance optimizations at this stage of the game.

**Comment by ahorner** [comment]
@danhodos: Ready for another look. I agree on the repetitive API requests, but I think the solution is most likely just to layer in some request caching later on, which shouldn't require much mangling of this code.

**Pull request merged**